controllers/integration/integration_adapter.go

```diff
@@ -137,6 +137,136 @@ func (a *Adapter) getAllApplicationSnapshots() (*[]appstudioshared.ApplicationSn
+	if err != nil {
+		return nil, err
+	}
+
+	return &applicationSnapshotEnvironmentBindingList.Items[0], nil
+}
+
+func (a *Adapter) EnsureApplicationSnapshotEnvironmentBindingExist() (results.OperationResult, error) {
+	availableEnvironments, err := a.findAvailabeEnvironments()
+	if err != nil {
+		return results.RequeueWithError(err)
+	}
+
+	applicationSnapshot, err := a.findMatchingApplicationSnapshot()
+	if err != nil {
+		return results.RequeueWithError(err)
+	}
+
+	components, err := a.getAllApplicationComponents(a.application)
+
+	if err != nil {
+		return results.RequeueWithError(err)
+	}
+	for _, availableEnvironment := range *availableEnvironments {
+		bindingName := a.application.Name + "-" + availableEnvironment.Spec.DisplayName + "-" + "binding"
+		applicationSnapshotEnvironmentBinding, _ := a.findExistingApplicationSnapshotEnvironmentBinding(availableEnvironment.Spec.DisplayName)
+		if applicationSnapshotEnvironmentBinding != nil {
+			applicationSnapshotEnvironmentBinding.Spec.Snapshot = applicationSnapshot.Spec.DisplayName
+			applicationSnapshotComponents := a.getApplicationSnapshotComponents(*components)
+			applicationSnapshotEnvironmentBinding.Spec.Components = *applicationSnapshotComponents
+		} else {
+			applicationSnapshotEnvironmentBinding, err := a.CreateApplicationSnapshotEnvironmentBinding(
+				bindingName, a.application.Namespace, a.application.Name,
+				availableEnvironment.Spec.DisplayName,
+				applicationSnapshot, *components)
+
+			if err != nil {
+				a.logger.Error(err, "Failed to create ApplicationSnapshotEnvironmentBinding",
+					"ApplicationSnapshotEnvironmentBindingName.Application", applicationSnapshotEnvironmentBinding.Spec.Application,
+					"ApplicationSnapshotEnvironmentBindingName.Environment", applicationSnapshotEnvironmentBinding.Spec.Environment,
+					"ApplicationSnapshotEnvironmentBindingName.Snapshot", applicationSnapshotEnvironmentBinding.Spec.Snapshot)
+				return results.RequeueOnErrorOrStop(a.updateStatus())
+			}
+
+			a.logger.Info("Created new ApplicationSnapshotEnvironmentBinding",
+				"ApplicationSnapshotEnvironmentBindingName.Application", applicationSnapshotEnvironmentBinding.Spec.Application,
+				"ApplicationSnapshotEnvironmentBindingName.Environment", applicationSnapshotEnvironmentBinding.Spec.Environment,
+				"ApplicationSnapshotEnvironmentBindingName.Snapshot", applicationSnapshotEnvironmentBinding.Spec.Snapshot)
+		}
+	}
+	return results.ContinueProcessing()
+}
+
+func (a *Adapter) getApplicationSnapshotComponents(components []hasv1alpha1.Component) *[]appstudioshared.BindingComponent {
+	bindingComponents := []appstudioshared.BindingComponent{}
+	for _, component := range components {
+		bindingComponents = append(bindingComponents, appstudioshared.BindingComponent{
+			Name: component.Spec.ComponentName,
+			Configuration: appstudioshared.BindingComponentConfiguration{
+				Env: []appstudioshared.EnvVarPair{
```

Comment:
We don't need to set the `Env` here as it's field that's set by the end-users. Also the component doesn't have to have `Env` variables defined in its `Spec`, so we would run into an error here in that case.

Reply:
Removed Env part. Please help to review again. Thanks